### DISCUSSION TITLE
Textadept 12.0 alpha 2

### DISCUSSION THREAD

**Discussion by orbitalquark** [orbitalquark/textadept#368] [General]
Textadept 12.0 alpha 2 is released from https://github.com/orbitalquark/textadept/releases/tag/textadept_12.0_alpha_2

Like the previous alpha, this new alpha also contains major changes, such as:

  - Language modules are no longer auto-loaded.
  - The external LSP module has received a number of major improvements.
  - All previously included language modules are no longer included or maintained. (Their functionality has been supplanted by the LSP module and language servers. Note that that module contains its own basic Lua language server.)

Please see the [migration guide][] for migrating from Textadept 11 to 12. You can read more about the above changes [here][].

Bugfixes:

- Fixed auto-pair removal bug when backspacing over an auto-paired character.
- Fixed directory filters that contain directories to include.
- `textadept.editing.autocomplete()` should return `false` if no completions are displayed.
- Fixed clearing of a previous buffer's state when switching between buffers after closing one.
- Qt version: ensure the given directory is selected in Linux's directory selection dialog.
- Terminal version: fixed spawning processes on Linux and macOS.
- Terminal version: synchronize paragraph up/down extend selection key binding with GUI version.
- Scintilla: fixed scroll position update after `view:vertical_center_caret()`.
- Scintilla: prevent autocompletion of lists created during `events.CHAR_ADDED`.
- LSP: fixed diagnostic display.
- LSP: silently print to LSP buffer when starting language servers.
- LSP: properly handle incoming server requests.
- LSP: fixed "Find References".
- LSP: gracefully shut down language servers on quit too, not just reset.
- LSP: improved startup and error notifications.
- LSP: synchronize unsaved document changes when switching away from a buffer or view.
- LSP: handle unexpected language server exit (i.e. crash).
- LSP: fixed display of signature help when more than one option is present.

Changes:

- Language modules are no longer auto-loaded when their files are opened.
- All language autocompletion and documentation has been delegated to the external LSP module.
  - Deprecated C, CSS, Go, HTML, Lua, Python, and Ruby language modules.
  - Textadept no longer generates or uses Lua tags and api documentation files.
  - Deprecated the ctags module.
  - Removed "Tools > Complete Symbol" and "Tools > Show Documentation" menu items.
  - Removed `textadept.editing.api_files` and `textadept.editing.show_documentation()`.
- Hide the Find & Replace pane after an "In Files" search.
- Allow lexers to dictate what constitutes a word character.
- Added [`lexer.names()`][].
- Removed `textadept.editing.transpose_chars()`.
- Replaced `textadept.editing.typeover_chars` with [`typeover_auto_paired`][].
- Renamed `textadept.snippets.cancel_current()` to `cancel()`.
- Removed `textadept.snippets.path`.
- Added buffer representation argment to `events.BUFFER_DELETED`.
- Switched documentation format to [LDoc][] from LuaDoc.
- Added `-L` and `--lua` command line option for running Textadept as a standalone Lua interpreter.
- Scroll all views showing print/output buffers when printed to.
- Renamed `_SCINTILLA.next_*` to `_SCINTILLA.new_*`.
- `textadept.run.run_in_background` applies even if the output buffer is not open.
- Notify of compile/run/build/test/project command success or failure in statusbar.
- Autoscroll to the bottom of compile/run/build/test/project output buffer if possible.
- Removed `ui.command_entry.append_history()` and `ui.command_entry.height`.
- Use Scintillua as a Lua library instead of as a Scintilla lexer.
- Removed support for `buffer.property_int` (not `lexer.property_int`).
- LSP: updated to LSP 3.17.
- LSP: implemented selection range.
- LSP: support language-specific completion and signature help trigger characters.
- LSP: support highlighting active parameters in signature help.
- LSP: show relative paths in "Go To ..." dialogs if possible.
- LSP: send "textDocument/didClose" notifications.
- LSP: added simple Lua language server and enable it by default for Lua files.
- LSP: stop logging to a buffer and added "Show Log" menu option instead.
- LSP: allow for launching servers outside a project.

[migration guide]: https://orbitalquark.github.io/textadept/manual.html#migrating-from-textadept-11-to-12
[here]: https://github.com/orbitalquark/textadept/discussions/362
[`lexer.names()`]: https://orbitalquark.github.io/textadept/api.html#lexer.names
[`typeover_auto_paired`]: https://orbitalquark.github.io/textadept/api.html#textadept.editing.typeover_auto_paired
[LDoc]: https://stevedonovan.github.io/ldoc/

**Comment by paaguti** [comment]
> Removed textadept.snippets.path.

May I ask why? The mix of shorter snippets in the Lua code and longer snippets in files is so convenient it would be a pity to loose that. One thing is having an LSP to assist in the creation of code and the other is have snippets with things that you do repetitively. And these things might be large and more convenient to do verbatim in a file than with escapes, etc in Lua code.
For the sake of an example, Emacs et.al. are introducing LSP et.al. but no one has killed/retired snippets, templates, or however you want to call them.

**Comment by orbitalquark** [comment]
My thinking was that one can use *~/.textadept/init.lua* to import 
snippets from an arbitrary directory structure.

**Reply by paaguti** [reply]
Maybe it would be good to include some sample code for newbies... I'm using my hydra-based snippets infrastructure, but other people may want a simpler solution ;-)

**Comment by orbitalquark** [comment]
Fair enough. Here's a new wiki page linked from the wiki home page in 
the snippets section: 
https://github.com/orbitalquark/textadept/wiki/Load-snippets-in-directories

**Reply by paaguti** [reply]
After playing with this for some weeks, I feel that it would be nice to settle down for one directory for snippets, with a default value (for example `$HOME/.textadept/snippets`) were snippets could reside and putting a reduced version of the WIKI code back to the TextAdept source. 
Of course, if a user wants a full multi-directory structure, there would be the code in the Wiki to extend TA's behaviour.

IMvvHO, snippets are too cool a feature to 'hide' it from new users by making them find some code in the Wiki and putting that code in init.lua. My gut feeling is that I'd like them sort of 'out-of-the-box'. That's what 'dragged' me :-) into TA.

**Reply by orbitalquark** [reply]
Very well. `textadept.snippets.path` has been restored: https://github.com/orbitalquark/textadept/commit/bdcaf0328001247f6313f2ca444521ef43e52915

**Comment by paaguti** [comment]
Cool!Enviado desde mi iPhoneEl 5 mar 2023, a las 16:12, orbitalquark ***@***.***> escribió:﻿
Fair enough. Here's a new wiki page linked from the wiki home page in 
the snippets section: 
https://github.com/orbitalquark/textadept/wiki/Load-snippets-in-directories

**Comment by ghost** [comment]
Please have a look at lsp over tcp: https://github.com/orbitalquark/textadept-lsp/issues/12

**Comment by ghost** [comment]
After your fix for https://github.com/orbitalquark/textadept/discussions/362#discussioncomment-5069003 (I'm not really sure if it's the cause but I discovered it's this way since this 12.0 alpha 2), the editor will only start to highlight after I stopped typing. This caused an impression that the editor is slow as there is a delay.

**Reply by orbitalquark** [reply]
I don't think it was that change, but I did revert something that may have caused what you're seeing: https://github.com/orbitalquark/textadept/commit/995b00999872f09d1364daf6951830cb166925ac. It's in the latest nightly build. Let me know if that did not fix it.

**Reply by ghost** [reply]
It seems the problem was fixed on the latest nightly.

**Comment by ghost** [comment]
Should the Open Recent... dialog be extended to support selecting of multiple files to open? On other text editors I know, they show the recently opened files directly via a sub menu with a "Clear recently opened files" entry at the end to clear recently opened files. TextAdept implemented it differently so there must be a reason to do so. I think it's natural to extend the Open Recent... dialog to support selecting of multiple files to open. Otherwise, it's just a weird decision imo. The Open Recent... dialog in it current form is just inconvenient to work with. Btw, please add "Clear recently opened files" to the Open Recent... dialog. Thank you.

**Reply by orbitalquark** [reply]
I've added a new issue for this: #373 . Please open new issues or discussions instead of adding to this thread unless you are talking about features specifically added, changed, or removed from 12.0 alpha 2.

**Comment by ghost** [comment]
I just tried to build TextAdept from source. Given the amount of the dependencies, I don't think adding https://github.com/itay-grudev/SingleApplication to support single instance on Qt is something unreasonable. Please reconsider: https://github.com/orbitalquark/textadept/issues/361

**Reply by orbitalquark** [reply]
I re-opened it.

**Comment by ghost** [comment]
Btw, you always said you can't implement theming for TextAdept (I mean the TextAdept app, not the editor color scheme you called theme). You always said it's because the limitations of GTK and Qt so you can't do anything with it. The fact is, there is no such limitations. Both GTK and Qt support style sheets:

https://docs.gtk.org/gtk3/css-overview.html

https://doc.qt.io/qt-5/stylesheet-customizing.html

An example of a Qt app with out of the box dark theme: https://github.com/royqh1979/RedPanda-CPP

**Reply by orbitalquark** [reply]
I've addressed this topic in #375.

**Reply by ghost** [reply]
Are you going to implement dark mode soon? You said you already have a reference on https://github.com/orbitalquark/textadept/issues/336

**Reply by orbitalquark** [reply]
Probably not. If you're on Linux or macOS, you can already set your Qt theme to a dark mode and use a dark Textadept theme. Only Windows doesn't change Textadept's widget theme when turning on dark mode, and that might be a Qt issue.

**Comment by jxzwp** [comment]
Please consider re-adding `ui.command_entry.height` and the `height` parameter to `ui.command_entry.run()`, which were removed in https://github.com/orbitalquark/textadept/commit/611d7c80a20a67822979490fb790c5cbaf76aee6

Allowing the height to be changed enables the command entry buffer to be reused for many other purposes.

For instance prior to TA 12 I was sending run output to it, so it functioned as a transient pop-up output buffer.
The advantage of this was I could rapidly perform "edit -> run -> view-output" cycles: 
+ without wasting any screen space on a dedicated view for the [Output Buffer].
+ I didn't need to worry about how I'd previously created the split table of views and whether the view displaying the [Output Buffer] would be too narrow to be useful.
+ I could hide the output and return to the executed file with a single keypress (escape).
+ Because I'd clear the previous output before each run, I'd never accidentally look at old output and draw the wrong conclusion about the effects of the edit I'd just made.

The command entry is well suited to display information rapidly on demand, like code metrics, git status or api docs with syntax-highlighting and a larger font size than is practical via a call tip.

I don't think I'm alone in repurposing the command entry, the Textredux module has a method for attaching a reduxbuffer to the command entry https://github.com/rgieseke/textredux/blob/b7900c77e1604ed0f0c3a4b9c0b0776e89bb7b2a/core/buffer.lua#L244

With the command entry limited to a height of one line in TA 12 alpha 2, all of the above becomes impractical, because the user either has to view the content one line at a time, by repeatedly pressing the down arrow; or use the mouse to make the view larger on each and every use.

Thanks for creating Textadept, I appreciate all the time you've put into it. I've been a user since version 9.2, and have purchased multiple versions of the quick reference over the years.

**Reply by orbitalquark** [reply]
Thanks for your very thoughtful feedback. I will look into restoring that field.

**Reply by orbitalquark** [reply]
`ui.command_entry.height` has been restored and will be in the next nightly build: https://github.com/orbitalquark/textadept/commit/95104a627a2c3c18176745a543b743a6986a6992

**Comment by paaguti** [comment]
Greatly appreciated :-)

On Fri, 7 Apr 2023 at 00:51, orbitalquark ***@***.***> wrote:

> Very well. textadept.snippets.path has been restored: bdcaf03
> <https://github.com/orbitalquark/textadept/commit/bdcaf0328001247f6313f2ca444521ef43e52915>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orbitalquark/textadept/discussions/368#discussioncomment-5548592>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZKU4CNHHGO43AI6KK4WALW75CHBANCNFSM6AAAAAAVMVYC2A>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Fragen sind nicht da, um beantwortet zu werden,
Fragen sind da um gestellt zu werden
Georg Kreisler

Headaches with a Juju log:
unit-basic-16: 09:17:36 WARNING juju.worker.uniter.operation we should run
a leader-deposed hook here, but we can't yet
